BlockObjectToStateSerializer: Avoid unnecessary Writer and Closure

ShockedPlot7560 opened the pull request:
In the case of a simple mapping for the block serializer, we don't even need to go through `BlockStateWriter` to get the `BlockStateData`.

This PR, as the issue suggests, allows the serializer to return a string directly instead of a Writer in order to limit the allocation of unnecessary objects.

### Related issues & PRs
* fixes #6675 

## Changes
### API changes
- Serializers in `BlockObjectToStateSerializer` can now return a string safely.

ShockedPlot7560 committed:
BlockObjectToStateSerializer: Allow serializers to return an ID string

ShockedPlot7560 committed:
fix PHPstan

dktapps commented:
I did look at doing this myself, but since we already cache state IDs in this case anyway, the benefit is fairly minimal. It does allow us to have less useless closures and such, but the performance benefit is close to 0.

Edit: I see this goes further than I originally envisioned. I hadn't thought about getting rid of `Writer` for switched IDs with no states.

dktapps commented on src/data/bedrock/block/convert/BlockObjectToStateSerializer.php:
```diff
@@ -243,7 +243,7 @@ public function map(Block $block, \Closure $serializer) : void{
 
 	public function mapSimple(Block $block, string $id) : void{
-		$this->map($block, fn() => Writer::create($id));
+		$this->map($block, fn() => $id);
```
This doesn't make much sense, it would be better if we could ditch the closure entirely

ShockedPlot7560 replied:
Yeah, that's was my second intention, it will be done in this PR instead of a follow up

ShockedPlot7560 committed:
Get rid of useless closure

ShockedPlot7560 committed:
Fix style

dktapps commented on src/data/bedrock/block/convert/BlockObjectToStateSerializer.php:
```diff
@@ -277,14 +278,20 @@ public function serializeBlock(Block $blockState) : BlockStateData{
+		 * @phpstan-var \Closure(TBlockType) : (Writer|string)|string $serializer
 		 */
 		$serializer = $locatedSerializer;
+		if(is_callable($serializer)){
```
`instanceof \Closure` would be better here, a string ID might be incorrectly considered callable

dktapps replied:
This can also be reduced to a ternary

dktapps replied:
Actually this whole thing can be simplified

ShockedPlot7560 committed:
Simplify serializeBlock

dktapps commented:
Are there any situations where we return a `Writer` without taking any info from the `Block`?

ShockedPlot7560 commented:
> Are there any situations where we return a `Writer` without taking any info from the `Block`?

I've gone through everything and normally there shouldn't be any more, subject to error but low risk.

dktapps commented:
The code you've written would allow for `Writer|string` or `Closure(Block) : (Writer|string)` if we wanted it to. So I'm just wondering if there's any cases where we write properties that aren't derived from the block (e.g. to polyfill and make Minecraft happy).

dktapps commented:
https://github.com/pmmp/PocketMine-MP/blob/stable/src/data/bedrock/block/convert/BlockObjectToStateSerializer.php#L594
https://github.com/pmmp/PocketMine-MP/blob/a83c62a4a2e5eea17fdfc3eb2bfd55f31534a9e3/src/data/bedrock/block/convert/BlockObjectToStateSerializer.php#L1282-L1283
https://github.com/pmmp/PocketMine-MP/blob/a83c62a4a2e5eea17fdfc3eb2bfd55f31534a9e3/src/data/bedrock/block/convert/BlockObjectToStateSerializer.php#L1629-L1630
https://github.com/pmmp/PocketMine-MP/blob/a83c62a4a2e5eea17fdfc3eb2bfd55f31534a9e3/src/data/bedrock/block/convert/BlockObjectToStateSerializer.php#L1706
https://github.com/pmmp/PocketMine-MP/blob/a83c62a4a2e5eea17fdfc3eb2bfd55f31534a9e3/src/data/bedrock/block/convert/BlockObjectToStateSerializer.php#L1714

All of these cases could be mapped to a `Writer` instance directly (or `BlockStateData`) since they don't use the block input for context. If you ctrl+f for `fn() =>` there's probably more cases.

ShockedPlot7560 commented:
I see our point of view, it is necessary to define the serialiser as `Closure|string|Writer` to cover all cases. Good catch

dktapps commented:
I've been thinking about this, and I'm not sure that going to `string` actually makes sense. Since we turn it into `BlockStateData` before returning it anyway, and there are cases where we just want to return a static `BlockStateData` anyway.

Perhaps it would make more sense to just have `BlockStateData|\Closure(Block) : (Writer|BlockStateData)`.

ShockedPlot7560 committed:
Use of BlockStateData instead of string

I'm not sure at the moment about using BlockStateData::current and getBlockStateData directly in this class.

ShockedPlot7560 commented:
I'm not sure at the moment about using BlockStateData::current and getBlockStateData directly in this class. Looks weird for me.

I'm still dubious about the use of BlockStateData rather string in the sense that the changes there are aimed more at reducing the memory footprint than overall performance. By replacing Writer with BlockStateData, there's no real gain, so do we really need this change?

dktapps committed:
tidy

dktapps committed:
Allow providing a plain Writer

dktapps committed:
tidy

dktapps committed:
avoid specialising for uncommon case

dktapps committed:
missed one

dktapps committed:
ffs

dktapps committed:
tidy

dktapps commented:
I've tidied this up. There are now the following cases:

- Closure(Block) -> Writer (regular case)
- Closure(Block) -> BlockStateData (block with flattened IDs, no state data)
- Writer (polyfills useless properties, not dependent on input)
- BlockStateData (ID only, not dependent on input)

Though I want to rethink the way flattened IDs are handled at some point
